File: go/mechanisms/evm/exact/facilitator/erc4337_bundler.go
package facilitator

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"time"
)

// BundlerClient is a JSON-RPC client for ERC-4337 bundler operations.
type BundlerClient struct {
	rpcUrl string
	config BundlerClientConfig
	client *http.Client
}

// NewBundlerClient creates a new BundlerClient.
func NewBundlerClient(rpcUrl string, config *BundlerClientConfig) *BundlerClient {
	cfg := BundlerClientConfig{
		TimeoutMs: 10000,
		Retries:   0,
	}
	if config != nil {
		if config.TimeoutMs > 0 {
			cfg.TimeoutMs = config.TimeoutMs
		}
		if config.Retries > 0 {
			cfg.Retries = config.Retries
		}
	}

	return &BundlerClient{
		rpcUrl: rpcUrl,
		config: cfg,
		client: &http.Client{},
	}
}

// EstimateUserOperationGas estimates gas for a user operation.
func (b *BundlerClient) EstimateUserOperationGas(ctx context.Context, userOp interface{}, entryPoint string) (*GasEstimate, error) {
	result, err := b.call(ctx, "eth_estimateUserOperationGas", []interface{}{userOp, entryPoint})
	if err != nil {
		return nil, err
	}

	data, err := json.Marshal(result)
	if err != nil {
		return nil, &BundlerError{
			Message:    fmt.Sprintf("failed to marshal gas estimate: %s", err.Error()),
			Method:     "eth_estimateUserOperationGas",
			BundlerUrl: b.rpcUrl,
		}
	}

	var gasEstimate GasEstimate
	if err := json.Unmarshal(data, &gasEstimate); err != nil {
		return nil, &BundlerError{
			Message:    fmt.Sprintf("failed to parse gas estimate: %s", err.Error()),
			Method:     "eth_estimateUserOperationGas",
			BundlerUrl: b.rpcUrl,
		}
	}

	return &gasEstimate, nil
}

// SendUserOperation sends a user operation to the bundler.
func (b *BundlerClient) SendUserOperation(ctx context.Context, userOp interface{}, entryPoint string) (string, error) {
	result, err := b.call(ctx, "eth_sendUserOperation", []interface{}{userOp, entryPoint})
	if err != nil {
		return "", err
	}

	hash, ok := result.(string)
	if !ok {
		return "", &BundlerError{
			Message:    "unexpected result type from eth_sendUserOperation",
			Method:     "eth_sendUserOperation",
			BundlerUrl: b.rpcUrl,
		}
	}

	return hash, nil
}

// GetUserOperationReceipt gets the receipt for a user operation.
func (b *BundlerClient) GetUserOperationReceipt(ctx context.Context, userOpHash string) (*UserOperationReceipt, error) {
	result, err := b.call(ctx, "eth_getUserOperationReceipt", []interface{}{userOpHash})
	if err != nil {
		return nil, err
	}

	if result == nil {
		return nil, nil
	}

	data, err := json.Marshal(result)
	if err != nil {
		return nil, &BundlerError{
			Message:    fmt.Sprintf("failed to marshal receipt: %s", err.Error()),
			Method:     "eth_getUserOperationReceipt",
			BundlerUrl: b.rpcUrl,
		}
	}

	var receipt UserOperationReceipt
	if err := json.Unmarshal(data, &receipt); err != nil {
		return nil, &BundlerError{
			Message:    fmt.Sprintf("failed to parse receipt: %s", err.Error()),
			Method:     "eth_getUserOperationReceipt",
			BundlerUrl: b.rpcUrl,
		}
	}

	return &receipt, nil
}

// call makes a JSON-RPC call with timeout and retry.
func (b *BundlerClient) call(ctx context.Context, method string, params []interface{}) (interface{}, error) {
	reqBody := jsonRpcRequest{
		Jsonrpc: "2.0",
		ID:      1,
		Method:  method,
		Params:  params,
	}

	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, &BundlerError{
			Message:    fmt.Sprintf("failed to marshal request: %s", err.Error()),
			Method:     method,
			BundlerUrl: b.rpcUrl,
		}
	}

	maxAttempts := b.config.Retries + 1
	var lastErr error

	for attempt := 0; attempt < maxAttempts; attempt++ {
		timeoutCtx, cancel := context.WithTimeout(ctx, time.Duration(b.config.TimeoutMs)*time.Millisecond)

		req, err := http.NewRequestWithContext(timeoutCtx, "POST", b.rpcUrl, bytes.NewReader(bodyBytes))
		if err != nil {
			cancel()
			return nil, &BundlerError{
				Message:    fmt.Sprintf("failed to create request: %s", err.Error()),
				Method:     method,
				BundlerUrl: b.rpcUrl,
			}
		}
		req.Header.Set("Content-Type", "application/json")

		resp, err := b.client.Do(req)
		if err != nil {
			cancel()
			lastErr = err
			if ctx.Err() != nil {
				return nil, &BundlerError{
					Message:    fmt.Sprintf("bundler request timeout after %dms", b.config.TimeoutMs),
					Method:     method,
					BundlerUrl: b.rpcUrl,
				}
			}
			if attempt < maxAttempts-1 {
				time.Sleep(time.Duration(math.Pow(2, float64(attempt))*100) * time.Millisecond)
				continue
			}
			return nil, &BundlerError{
				Message:    fmt.Sprintf("bundler request failed: %s", lastErr.Error()),
				Method:     method,
				BundlerUrl: b.rpcUrl,
			}
		}

		respBody, err := io.ReadAll(resp.Body)
		resp.Body.Close()
		cancel()

		if err != nil {
			lastErr = err
			if attempt < maxAttempts-1 {
				time.Sleep(time.Duration(math.Pow(2, float64(attempt))*100) * time.Millisecond)
				continue
			}
			return nil, &BundlerError{
				Message:    fmt.Sprintf("failed to read response: %s", lastErr.Error()),
				Method:     method,
				BundlerUrl: b.rpcUrl,
			}
		}

		if resp.StatusCode != http.StatusOK {
			return nil, &BundlerError{
				Message:    fmt.Sprintf("bundler HTTP error: %d %s", resp.StatusCode, resp.Status),
				Method:     method,
				BundlerUrl: b.rpcUrl,
			}
		}

		var rpcResp jsonRpcResponse
		if err := json.Unmarshal(respBody, &rpcResp); err != nil {
			return nil, &BundlerError{
				Message:    fmt.Sprintf("failed to parse JSON-RPC response: %s", err.Error()),
				Method:     method,
				BundlerUrl: b.rpcUrl,
			}
		}

		if rpcResp.Error != nil {
			msg := "Bundler RPC error"
			if rpcResp.Error.Message != "" {
				msg = rpcResp.Error.Message
			}
			return nil, &BundlerError{
				Message:    msg,
				Code:       rpcResp.Error.Code,
				Data:       rpcResp.Error.Data,
				Method:     method,
				BundlerUrl: b.rpcUrl,
			}
		}

		return rpcResp.Result, nil
	}

	return nil, &BundlerError{
		Message:    "bundler request failed after retries",
		Method:     method,
		BundlerUrl: b.rpcUrl,
	}
}

File: python/x402/mechanisms/evm/exact/erc4337_bundler.py
"""ERC-4337 Bundler JSON-RPC client for x402 EVM mechanism."""

import json
import time
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from typing import Any


class BundlerError(Exception):
    """Custom error for bundler-related failures."""

    def __init__(
        self,
        message: str,
        code: int | None = None,
        data: Any = None,
        method: str | None = None,
        bundler_url: str | None = None,
    ):
        super().__init__(message)
        self.code = code
        self.data = data
        self.method = method
        self.bundler_url = bundler_url


@dataclass
class BundlerClientConfig:
    """Configuration for BundlerClient."""

    timeout: int = 10  # seconds
    retries: int = 0


@dataclass
class GasEstimate:
    """Gas estimation response from bundler."""

    call_gas_limit: str | None = None
    verification_gas_limit: str | None = None
    pre_verification_gas: str | None = None
    max_fee_per_gas: str | None = None
    max_priority_fee_per_gas: str | None = None
    paymaster_verification_gas_limit: str | None = None
    paymaster_post_op_gas_limit: str | None = None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "GasEstimate":
        return cls(
            call_gas_limit=data.get("callGasLimit"),
            verification_gas_limit=data.get("verificationGasLimit"),
            pre_verification_gas=data.get("preVerificationGas"),
            max_fee_per_gas=data.get("maxFeePerGas"),
            max_priority_fee_per_gas=data.get("maxPriorityFeePerGas"),
            paymaster_verification_gas_limit=data.get("paymasterVerificationGasLimit"),
            paymaster_post_op_gas_limit=data.get("paymasterPostOpGasLimit"),
        )


@dataclass
class UserOperationReceipt:
    """User operation receipt from bundler."""

    user_op_hash: str
    entry_point: str
    sender: str
    nonce: str
    actual_gas_cost: str
    actual_gas_used: str
    success: bool
    reason: str | None = None
    logs: list[Any] = field(default_factory=list)
    transaction_hash: str | None = None
    receipt_transaction_hash: str | None = None

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "UserOperationReceipt":
        receipt = data.get("receipt", {})
        return cls(
            user_op_hash=data.get("userOpHash", ""),
            entry_point=data.get("entryPoint", ""),
            sender=data.get("sender", ""),
            nonce=data.get("nonce", ""),
            actual_gas_cost=data.get("actualGasCost", ""),
            actual_gas_used=data.get("actualGasUsed", ""),
            success=data.get("success", False),
            reason=data.get("reason"),
            logs=data.get("logs", []),
            transaction_hash=data.get("transactionHash"),
            receipt_transaction_hash=receipt.get("transactionHash")
            if isinstance(receipt, dict)
            else None,
        )


class BundlerClient:
    """JSON-RPC client for ERC-4337 bundler operations."""

    def __init__(self, rpc_url: str, config: BundlerClientConfig | None = None):
        self._rpc_url = rpc_url
        self._config = config or BundlerClientConfig()

    def estimate_user_operation_gas(self, user_op: dict[str, Any], entry_point: str) -> GasEstimate:
        """Estimate gas for a user operation."""
        result = self._call("eth_estimateUserOperationGas", [user_op, entry_point])
        return GasEstimate.from_dict(result)

    def send_user_operation(self, user_op: dict[str, Any], entry_point: str) -> str:
        """Send a user operation and return the hash."""
        result = self._call("eth_sendUserOperation", [user_op, entry_point])
        if not isinstance(result, str):
            raise BundlerError(
                "unexpected result type from eth_sendUserOperation",
                method="eth_sendUserOperation",
                bundler_url=self._rpc_url,
            )
        return result

    def get_user_operation_receipt(self, user_op_hash: str) -> UserOperationReceipt | None:
        """Get the receipt for a user operation."""
        result = self._call("eth_getUserOperationReceipt", [user_op_hash])
        if result is None:
            return None
        return UserOperationReceipt.from_dict(result)

    def _call(self, method: str, params: list[Any]) -> Any:
        """Make a JSON-RPC call with timeout and retry."""
        request_body = json.dumps(
            {
                "jsonrpc": "2.0",
                "id": 1,
                "method": method,
                "params": params,
            }
        ).encode("utf-8")

        max_attempts = self._config.retries + 1

        for attempt in range(max_attempts):
            try:
                req = urllib.request.Request(
                    self._rpc_url,
                    data=request_body,
                    headers={"Content-Type": "application/json"},
                    method="POST",
                )

                with urllib.request.urlopen(req, timeout=self._config.timeout) as resp:
                    if resp.status != 200:
                        raise BundlerError(
                            f"Bundler HTTP error: {resp.status}",
                            method=method,
                            bundler_url=self._rpc_url,
                        )

                    response_data = json.loads(resp.read().decode("utf-8"))

                if "error" in response_data and response_data["error"] is not None:
                    error = response_data["error"]
                    raise BundlerError(
                        error.get("message", "Bundler RPC error"),
                        code=error.get("code"),
                        data=error.get("data"),
                        method=method,
                        bundler_url=self._rpc_url,
                    )

                return response_data.get("result")

            except BundlerError:
                raise
            except urllib.error.URLError as e:
                if attempt < max_attempts - 1:
                    time.sleep(2**attempt * 0.1)
                    continue
                raise BundlerError(
                    f"Bundler request failed: {e}",
                    method=method,
                    bundler_url=self._rpc_url,
                ) from e
            except Exception as e:
                if attempt < max_attempts - 1:
                    time.sleep(2**attempt * 0.1)
                    continue
                raise BundlerError(
                    f"Bundler request failed: {e}",
                    method=method,
                    bundler_url=self._rpc_url,
                ) from e

        raise BundlerError(
            "Bundler request failed after retries",
            method=method,
            bundler_url=self._rpc_url,
        )
